src/dist/dist.rs

```diff
@@ -599,6 +600,87 @@ fn update_from_dist_<'a>(
     profile: Option<Profile>,
     prefix: &InstallPrefix,
     force_update: bool,
+) -> Result<Option<String>> {
+    let mut toolchain = toolchain.clone();
+    let mut fetched = String::new();
+    let mut first_err = None;
+    let backtrack = toolchain.channel == "nightly" && toolchain.date.is_none();
+    loop {
+        match try_update_from_dist_(
+            download,
+            update_hash,
+            &toolchain,
+            profile,
+            prefix,
+            force_update,
+            &mut fetched,
+        ) {
+            Ok(v) => break Ok(v),
+            Err(e) => {
+                if !backtrack {
+                    break Err(e);
+                }
+
+                if let ErrorKind::RequestedComponentsUnavailable(components, ..) = e.kind() {
+                    (download.notify_handler)(Notification::SkippingNightlyMissingComponent(
+                        components,
+                    ));
+
+                    if first_err.is_none() {
+                        first_err = Some(e);
+                    }
+                } else if let Some(e) = first_err {
+                    // if we fail to find a suitable nightly, we abort the search and give the
+                    // original "components unavailable for download" error.
+                    break Err(e);
+                } else {
+                    break Err(e);
+                }
+
+                // The user asked to update their nightly, but the latest nightly does not have all
+                // the components that the user currently has installed. Let's try the previous
+                // nightly (recursively) until we find a nightly that does.
```

Comment:
Strictly speaking this is iterative not recursive, but I wouldn't block merge on that nitpick alone